data_parallel_tensor.py

```diff
@@ -0,0 +1,396 @@
+            requires_grad=meta_t.requires_grad,
+        )
+        r.elem = dpt
+        return r
+
+    def __repr__(self):
+        if self.grad_fn:
+            return f"DataParallelTensor({self.elem}, grad_fn={self.grad_fn})"
+        return f"DataParallelTensor({self.elem})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def wrap(e):
+            if isinstance(e, DataParallelTensor):
+                return e
+            elif isinstance(e, torch.Tensor):
+                return DataParallelTensor(e, func, DPTensorType.replicated)
+            else:
+                return e
+
+        # All the args and kwargs are checked and any leaf tensors are wrapped as replicated DPTs
+        args = tree_map(wrap, args)
+        kwargs = tree_map(wrap, kwargs)
+
+        def unwrap_with_position(pos):
+            def get_element(e):
+                return e.elem[pos] if isinstance(e, DataParallelTensor) else e
+
+            return get_element
+
+        # Call the function for each of the DPT elements by unwarpping them and corresponding args and kwargs,
+        #  into element tensors so that the operation is performed on all the elements residing on the same device
+        if PARALLEL_DISPATCH:
+            future_res: List[futures.Future] = []
+            for pos in range(cls.num_threads):
+                future_res.append(
+                    cls.threadpool.submit(
+                        func,
+                        *tree_map(unwrap_with_position(pos), args),
+                        **tree_map(unwrap_with_position(pos), kwargs),
+                    )
+                )
+            outs = [future_res[i].result() for i in range(cls.num_threads)]
+        else:
+            outs = []
+            for pos in range(len(cls.device_ids)):
+                outs.append(
+                    func(
+                        *tree_map(unwrap_with_position(pos), args),
+                        **tree_map(unwrap_with_position(pos), kwargs),
+                    )
+                )
+
+        def get_element_type(lis):
+            assert isinstance(lis, list)
+            return type(lis[0])
+
+        # The ouput will always be a list
+        # The list can contain tensors, bools, list of tensors or tuples of tensors or None
+        # In case of tensors we just wrap them in DPT
+        # In case of list/tuple of tensors, the corresponding elements across list/tuple are warpped
+        #  into a DPT and a list/tuple is returned respectively
+
+        def out_wrap(e, func):
+            elem_type = get_element_type(e)
+            if elem_type is NoneType:
+                return None
+            if elem_type == torch.Tensor:
+                return DataParallelTensor(outs, func, DPTensorType.distributed)
+            elif elem_type == list:
+                return list(
+                    DataParallelTensor(list(t), func, DPTensorType.distributed)
+                    for t in zip(*e)
```

Comment:
Can you have scenarios where there is a list of list of Tensors?

Reply:
No we do not have such scenarios, either a single tensor or a list/tuple of tensors is returned by all operators that return a tensor.